sw/device/tests/sim_dv/lc_ctrl_transition_test.c

```diff
@@ -0,0 +1,93 @@
+
+#include <assert.h>
+#include <stdbool.h>
+
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_lc_ctrl.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/check.h"
+#include "sw/device/lib/testing/test_main.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+static dif_lc_ctrl_t lc;
+
+const test_config_t kTestConfig;
+static const size_t kLcTokenSize = 128 / 8;
+
+// LC exit token value for LC state transition.
+static const uint8_t lc_exit_token[kLcTokenSize] = {
+    0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07,
+    0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f,
+};
+
+static void check_lc_state_transition_count(uint8_t exp_lc_count) {
+  LOG_INFO("Read LC count and check with expect_val=%0d", exp_lc_count);
+  uint8_t lc_count;
+  CHECK(dif_lc_ctrl_get_attempts(&lc, &lc_count) == kDifLcCtrlAttemptsOk,
+        "Read lc_count register failed!");
+  CHECK(lc_count == exp_lc_count,
+        "LC_count error, expected %0d but actual count is %0d", exp_lc_count,
+        lc_count);
+}
+
+/**
+ * Tests the state transition request handshake between LC_CTRL and OTP_CTRL.
+ *
+ * 1). OTP pre-load image with lc_count = `8`.
+ * 2). Backdoor write OTP's LC parition to `TestUnlocked2` state, and backdoor
+ * write OTP's `test_exit` token to match the pattern
+ * `h00_01_02_03_04_05_06_07_08_09_0a_0b_0c_0d`.
+ * 3). When LC_CTRL is ready, check LC_CNT and LC_STATE register.
+ * 4). Program LC state transition request to advance to `Prod` state.
+ * 5). Issue hard reset.
+ * 6). Wait for LC_CTRL is ready, then check if LC_STATE advanced to `Dev`
+ * state, and lc_count advanced to `9`.
+ */
+
+bool test_main(void) {
+  LOG_INFO("Start LC_CTRL transition test.");
+
+  mmio_region_t lc_reg = mmio_region_from_addr(TOP_EARLGREY_LC_CTRL_BASE_ADDR);
+  CHECK(dif_lc_ctrl_init((dif_lc_ctrl_params_t){.base_addr = lc_reg}, &lc) ==
+        kDifLcCtrlOk);
+
+  LOG_INFO("Read and check LC state.");
+  dif_lc_ctrl_state_t curr_state;
+  CHECK(dif_lc_ctrl_get_state(&lc, &curr_state) == kDifLcCtrlOk);
+
+  // The OTP preload image hardcode lc_count as 8.
+  const uint8_t LcStateTransitionCount = 8;
+  dif_lc_ctrl_state_t req_lc_ctrl_state = kDifLcCtrlStateDev;
+
+  if (curr_state == kDifLcCtrlStateTestUnlocked2) {
+    // LC TestUnlocked2 is the intial test state for this sequence.
+    // The sequence will check if lc_count matches the preload value.
+    check_lc_state_transition_count(LcStateTransitionCount);
+
+    // Request lc_state transfer to Dev state.
+    dif lc_token_t token;
+    token.data = lc_exit_token;
+    CHECK(dif_lc_ctrl_mutex_try_acquire(&lc) == kDifLcCtrlMutexOk);
+    CHECK(dif_lc_ctrl_transition(&lc, req_lc_ctrl_state, &token) ==
+              kDifLcCtrlMutexOk,
+          "LC_transition failed!");
+
+    LOG_INFO("Waiting for LC transtition done and reboot.");
+    wait_for_interrupt();
+  } else {
```

Comment:
super nit: add newline before this for clarity of clause separation

Reply:
Thanks for pointing that out. added!